src/main/java/carpet/patches/TickSyncedBorderExtent.java

```diff
@@ -0,0 +1,160 @@
+ * this class uses the in game ticks.
+ */
+@SuppressWarnings("JavadocReference")
+public class TickSyncedBorderExtent implements WorldBorder.BorderExtent
+{
+	private final WorldBorder border;
+	private final long realDuration;
+	private final double tickDuration;
+	private final double from;
+	private final double to;
+
+	private int ticks;
+
+	public TickSyncedBorderExtent(WorldBorder border, long realDuration, double from, double to)
+	{
+		this.border = border;
+		this.realDuration = realDuration;
+		this.tickDuration = realDuration / 50.0;
+		this.from = from;
+		this.to = to;
+		this.ticks = 0;
+	}
+
+	@Override
+	public double getMinX()
+	{
+		int maxSize = this.border.getAbsoluteMaxSize();
+		return Mth.clamp(this.border.getCenterX() - this.getSize() / 2.0, -maxSize, maxSize);
+	}
+
+	@Override
+	public double getMaxX()
+	{
+		int maxSize = this.border.getAbsoluteMaxSize();
+		return Mth.clamp(this.border.getCenterX() + this.getSize() / 2.0, -maxSize, maxSize);
+	}
+
+	@Override
+	public double getMinZ()
+	{
+		int maxSize = this.border.getAbsoluteMaxSize();
+		return Mth.clamp(this.border.getCenterZ() - this.getSize() / 2.0, -maxSize, maxSize);
+	}
+
+	@Override
+	public double getMaxZ()
+	{
+		int maxSize = this.border.getAbsoluteMaxSize();
+		return Mth.clamp(this.border.getCenterZ() + this.getSize() / 2.0, -maxSize, maxSize);
+	}
+
+	@Override
+	public double getSize()
+	{
+		double progress = this.ticks / this.tickDuration;
+		return progress < 1.0 ? Mth.lerp(progress, this.from, this.to) : this.to;
+	}
+
+	@Override
+	public double getLerpSpeed()
+	{
+		return Math.abs(this.from - this.to) / this.tickDuration;
+	}
+
+	@Override
+	public long getLerpRemainingTime()
+	{
+		// Rough estimation
+		MinecraftServer server = CarpetServer.minecraft_server;
+		double ms;
+		if (server == null) 
+		{
+		    ms = TickSpeed.mspt;
+		} 
+		else 
+		{
+		     OptionalDouble optional = Arrays.stream(server.tickTimes).average();
+		     // Optional should never be empty but just in case...
+		     ms = optional.isEmpty() ? TickSpeed.mspt : optional.getAsDouble() * 1.0E-6D;
+		}
+		double tps = 1_000.0D / Math.max((TickSpeed.time_warp_start_time != 0) ? 0.0 : TickSpeed.mspt, ms);
+		return (long) ((this.tickDuration - this.ticks) / tps * 1_000);
+	}
+
+	@Override
+	public double getLerpTarget()
+	{
+		return this.to;
+	}
+
+	@Override
+	public BorderStatus getStatus()
+	{
+		return this.to < this.from ? BorderStatus.SHRINKING : BorderStatus.GROWING;
+	}
+
+	@Override
+	public void onAbsoluteMaxSizeChange()
+	{
+
+	}
+
+	@Override
+	public void onCenterChange()
+	{
+
+	}
+
+	@Override
+	public WorldBorder.BorderExtent update()
+	{
+		if (this.ticks++ % 20 == 0)
+		{
+			for (BorderChangeListener listener : this.border.getListeners())
+			{
+				if (!(listener instanceof BorderChangeListener.DelegateBorderChangeListener))
+				{
+					listener.onBorderSizeLerping(this.border, this.from, this.to, this.realDuration);
+				}
+			}
```

Comment:
Can't find this in 1.19.3, may have changed since it was written. Saves an AW :)

Reply:
No, this is not part of the vanilla code. I added this to periodically update the client. This is needed as the client relies on the WB updating in real time. Real time doesn't change but the tick rate can change - for this reason we just update the lerping state every second.

You could probably implement this better by having it only update when the tick rate actually changes but I don't think that's worth while.

Reply:
It's also valid to update all listeners as we are assuming that they do not know about the `tick` update behaviour of this custom implementation so they would also need to be updated. The reason I exclude `DelegateBorderChangeListener` is because that shadows all changes to other world borders in different worlds, that is not wanted.